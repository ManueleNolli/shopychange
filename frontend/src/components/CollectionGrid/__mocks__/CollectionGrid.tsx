import { divWithoutChildrenMock } from '../../../../mocksUtils'

const CollectionGridMock = () => divWithoutChildrenMock('CollectionGrid')
export default CollectionGridMock
