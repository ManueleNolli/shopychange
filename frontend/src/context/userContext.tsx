import React, { createContext, useContext, useState, ReactNode } from 'react'

interface UserContextType {
  userAddress: string | null
  blockchainNetworkId: number | null
  updateUserContext: (address: string, network: number) => void
  clearUserContext: () => void
}

const UserContext = createContext<UserContextType | undefined>(undefined)

export function UserProvider({ children }: { children: ReactNode }) {
  const [userAddress, setUserAddress] = useState<string | null>(null)
  const [blockchainNetworkId, setBlockchainNetworkId] = useState<number | null>(
    null
  )

  const updateUserContext = (address: string, networkId: number) => {
    setUserAddress(address)
    setBlockchainNetworkId(networkId)
  }

  const clearUserContext = () => {
    setUserAddress(null)
    setBlockchainNetworkId(null)
  }

  return (
    <UserContext.Provider
      value={{
        userAddress,
        blockchainNetworkId,
        updateUserContext,
        clearUserContext,
      }}
    >
      {children}
    </UserContext.Provider>
  )
}

export function useUserContext(): UserContextType {
  const context = useContext(UserContext)
  if (!context) {
    throw new Error('useUserContext must be used within a UserProvider')
  }
  return context
}
