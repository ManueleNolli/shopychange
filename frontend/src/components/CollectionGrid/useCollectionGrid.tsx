import { useEffect, useState } from 'react'

import { Collection } from '../../types/components/Collection'
import { NFT } from '../../types/components/NFT'
import {
  contractQuery,
  handlerQueryErrorCatch,
} from '../../services/BackendService'
import { useUserContext } from '../../context/userContext'
import useFetching from '../../hooks/useFetching'

type useCollectionGridProps = {
  nfts: NFT[]
}

export default function useCollectionGrid({ nfts }: useCollectionGridProps) {
  const { userAddress, blockchainNetworkId } = useUserContext()
  const { isFetching, setIsFetching, isError, setIsError } = useFetching()
  const [collections, setCollections] = useState<Collection[]>([])

  useEffect(() => {
    // Group nfts by contract address
    const groupedNfts = nfts.reduce((acc: { [key: string]: NFT[] }, nft) => {
      const contractAddress = nft.contractAddress

      if (!acc[contractAddress]) {
        acc[contractAddress] = []
      }

      acc[contractAddress].push(nft)
      return acc
    }, {})

    // Get contract metadata for each contract address
    const contractAddresses = Object.keys(groupedNfts)

    const fetchContractMetadata = async (contractAddress: string) => {
      if (!userAddress || !blockchainNetworkId) return

      const response = await handlerQueryErrorCatch(
        contractQuery(blockchainNetworkId, contractAddress)
      )
      if (response.error) {
        setIsError(true)
        setIsFetching(false)
        return
      }

      const contract = response.result.contract
      return contract
    }

    const fetchContractsMetadata = async (contractAddresses: string[]) => {
      const contracts = await Promise.all(
        contractAddresses.map(fetchContractMetadata)
      )

      return contracts
    }

    fetchContractsMetadata(contractAddresses).then((contracts) => {
      for (const contract of contracts) {
        if (contract === undefined) {
          setIsError(true)
          setIsFetching(false)
          return
        }
        const nfts = groupedNfts[contract.address]

        const collection: Collection = {
          contract,
          nfts,
        }

        setCollections((collections) => [...collections, collection])
        setIsFetching(false)
      }
    })
  }, [])

  return { isFetching, isError, collections }
}
