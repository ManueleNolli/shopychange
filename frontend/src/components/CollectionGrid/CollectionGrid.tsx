import React from 'react'
import { Box, Link, Divider, Heading } from '@chakra-ui/react'
import { CollectionGridProps } from '../../types/props/CollectionGridProps'
import NFTCardGrid from '../NFTCardGrid/NFTCardGrid'
import { Link as LinkRouter } from 'react-router-dom'
import useCollectionGrid from './useCollectionGrid'
import QueryContainer from '../../utils/QueryContainer/QueryContainer'

export default function CollectionGrid({
  nfts,
  nftChildren,
  nftChildrenWithProps,
  saleButton,
}: CollectionGridProps) {
  const { isFetching, isError, collections } = useCollectionGrid({
    nfts,
  })

  return (
    <QueryContainer isLoading={isFetching} isError={isError}>
      {collections.map((collection) => (
        <Box key={collection.contract.address} m={4}>
          <Link as={LinkRouter} to={'/' + collection.contract.address}>
            <Heading size="md">{collection.contract.name}</Heading>
          </Link>

          <NFTCardGrid
            nfts={collection.nfts}
            nftChildrenWithProps={nftChildrenWithProps}
            nftChildren={nftChildren}
            saleButton={saleButton}
          />
          {/* divider is not needed for the last collection */}
          {collection !== collections[collections.length - 1] && <Divider />}
        </Box>
      ))}
    </QueryContainer>
  )
}
