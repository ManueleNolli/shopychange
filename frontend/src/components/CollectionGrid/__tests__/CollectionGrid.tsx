import React from 'react'
import CollectionGrid from '../CollectionGrid'

import { render, act } from '@testing-library/react'
import { useUserContext } from '../../../context/userContext'
import useCollectionGrid from '../useCollectionGrid'

jest.mock('../../../services/BackendService', () => ({
  handlerQuery: jest.fn(),
  contractQuery: jest.fn(),
}))

jest.mock('../../NFTCard/NFTCard')

jest.mock('../../../context/userContext', () => ({
  useUserContext: jest.fn(),
}))

const mockNavigate = jest.fn()
jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
}))

jest.mock('../useCollectionGrid', () => jest.fn())

const mockData = [
  {
    contractAddress: '0x0000000000',
    tokenId: 0,
    image:
      'https://ipfs.io/ipfs/bafybeietntip4sq3xmuwiz3s43yb3klrpa3okgtg5deysmbd5hbizhrb2q/0.png',
    name: 'Test NFT',
  },
  {
    contractAddress: '0x0000000000',
    tokenId: 1,
    image:
      'https://ipfs.io/ipfs/bafybeietntip4sq3xmuwiz3s43yb3klrpa3okgtg5deysmbd5hbizhrb2q/1.png',
    name: 'Test NFT 1',
  },
]

describe('CollectionGrid', () => {
  test('renders CollectionGrid component', async () => {
    ;(useUserContext as jest.Mock).mockReturnValueOnce({
      userAddress: '0x0000001',
      blockchainNetworkId: 1,
    })
    ;(useCollectionGrid as jest.Mock).mockReturnValue({
      isFetching: false,
      isError: false,
      collections: [
        {
          contract: {
            name: 'Test Collection',
            symbol: 'TEST',
            address: '0x0000000000',
          },
          nfts: mockData,
        },
      ],
    })

    const tree = render(<CollectionGrid nfts={mockData} />)
    await act(async () => {
      Promise.resolve()
    })

    expect(tree.container).toMatchSnapshot()
  })
})
