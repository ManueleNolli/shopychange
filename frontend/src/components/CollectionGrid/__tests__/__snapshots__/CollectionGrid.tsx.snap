// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionGrid renders CollectionGrid component 1`] = `
<div>
  <div
    class="css-1kxrbkv"
  >
    <a
      class="chakra-link css-0"
      to="/0x0000000000"
    >
      <h2
        class="chakra-heading css-0"
      >
        Test Collection
      </h2>
    </a>
    <div
      class="css-xbmy3v"
    >
      <div
        data-testid="NFTCard"
      />
      <div
        data-testid="NFTCard"
      />
    </div>
  </div>
</div>
`;
